athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/storage/StorageMetadata.java

```diff
@@ -0,0 +1,315 @@
+        for (String path : folderPaths) {
+            Optional<FieldValue> optionalFieldValue = FieldValue.from(path);
+            if (optionalFieldValue.isEmpty()) {
+                continue;
+            }
+            FieldValue fieldValue = optionalFieldValue.get();
+            Optional<EqualsExpression> optionalExpression = expressions.stream()
+                    .filter(expr -> expr.columnName().equalsIgnoreCase(fieldValue.getField()))
+                    .findFirst();
+            if (optionalExpression.isPresent()) {
+                LOGGER.info("Evaluating field value {} against the expression {}", fieldValue, expressions);
+                EqualsExpression expression = optionalExpression.get();
+                if (!expression.apply(fieldValue.getValue())) {
+                    return false;
+                }
+            }
+            else {
+                LOGGER.info("No expression found for field {}", fieldValue.getField());
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Builds the table schema based on the provided field
+     *
+     * @param table      Glue table object
+     * @return An instance of {@link Schema}
+     */
+    public Schema buildTableSchema(Table table, BufferAllocator allocator) throws URISyntaxException
+    {
+        SchemaBuilder schemaBuilder = SchemaBuilder.newBuilder();
+        String locationUri = table.getStorageDescriptor().getLocation();
+        URI storageLocation = new URI(locationUri);
+        List<Field> fieldList = getFields(storageLocation.getAuthority(), storageLocation.getPath(), table.getParameters().get(CLASSIFICATION_GLUE_TABLE_PARAM), allocator);
+
+        LOGGER.debug("Schema Fields\n{}", fieldList);
+        for (Field field : fieldList) {
+            if (isFieldTypeNull(field)) {
+                field = Field.nullable(field.getName().toLowerCase(), Types.MinorType.VARCHAR.getType());
+            }
+            else {
+                field = new Field(field.getName().toLowerCase(), new FieldType(field.isNullable(), field.getType(), field.getDictionary(), field.getMetadata()), field.getChildren());
+            }
+            schemaBuilder.addField(getCompatibleField(field));
+        }
+        return schemaBuilder.build();
+    }
+
+    private Field getCompatibleField(Field field)
+    {
+        String fieldName = field.getName().toLowerCase();
+        Types.MinorType fieldType = Types.getMinorTypeForArrowType(field.getType());
+        switch (fieldType) {
+            case TIMESTAMPNANO:
+            case TIMESTAMPSEC:
+            case TIMESTAMPMILLI:
+            case TIMEMICRO:
+            case TIMESTAMPMICRO:
+            case TIMENANO:
+                    return new Field(fieldName,
+                            new FieldType(field.isNullable(), Types.MinorType.DATEMILLI.getType(), field.getDictionary(),
+                                    field.getMetadata()), field.getChildren());
```

Comment:
I have already asked this multiple times and I still haven't received an answer. How does this work if the value is like 1000 nanoseconds and then you just change the field type to milliseconds?

1000 nanoseconds isn't the same thing as 1000 milliseconds.